text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,455 @@
+});
+wave.addCdkStage(new MyAppStage(...));
+```
+
+### Validations
+
+Waves and Stages can be decorated with `before` and `after` actions, intended for validations
+and checks.
+
+```ts
+const def = new cdkp.PipelineDefinition({
+  build: new cdkp.CdkBuild({ ... }),
+});
+def.addCdkStage(new MyAppStage('Beta', ...));
+
+const wave = def.addWave('Prod', {
+  before: [
+    new cdkp.ManualApprovalAction('Promote Beta to Prod'),
+  ],
+});
+
+wave.addCdkStage(new MyAppStage(...), {
+  after: [
+    new cdkp.ShellScriptAction('Validate', {
+      commands: [
+        'curl -Ssf https://mywebsite.com/ping',
+      ],
+    }),
+  ],
+});
+```
+
+### Using additional artifacts
+
+To clearly distinguish them from Stack Outputs, we will call artifacts "file sets". They
+can be consumed and produced by build actions and shell actions, and are always
+about whole directories:
+
+```ts
+const pipeline = new cdkp.PipelineDefinition({
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    additionalInputs: {
+      'subdir': cdkp.CodePipelineSource.gitHub('OWNER2/REPO2'),
+      '../side-dir': someStep,
+    },
+    commands: ['npm ci', 'npm run build'],
+    additionalOutputs: {
+      tests: cdkp.AdditionalOutput.fromDirectory('test'),
+    }
+  }),
+});
+
+pipeline.addCdkStage(stage, {
+  after: [
+    new cdkp.ShellScriptAction({
+      input: pipeline.additionalBuildOutput('tests'),
+      commands: ['node ./integ-tests'],
+      output: cdkp.AdditionalOutput.fromDirectory('./test-results'),
+    }),
+  ],
+});
+```
+
+### Using Stack Outputs
+
+Stack outputs are useful in validations:
+
+```ts
+const cdkStage = new MyAppStage(...);
+wave.addCdkStage(cdkStage, {
+  after: [
+    new cdkp.ShellScriptAction('Validate', {
+      environmentVariablesFromStackOutputs: {
+        ENDPOINT_URL: cdkp.StackOutput.fromCfnOutput(cdkStage.endpointUrl),
+      },
+      commands: [
+        'curl -Ssf $ENDPOINT_URL',
+      ],
+    }),
+  ],
+});
+```
+
+### CodeBuild-specific customizations
+
+AWS-specific customizations are passed as parameters to the Backend class for the well-defined
+CodeBuild projects (build, self-update, assets, ...)
+
+```ts
+const def = new cdkp.PipelineDefinition({
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    commands: ['npm ci', 'npm run build', 'npx cdk synth'],
+    environment: {
+      NPM_CONFIG_UNSAFE_PERM: 'true',
+    },
+  }),
+});
+def.addCdkStage(...);
+def.addCdkStage(...);
+
+new cdkp.CodePipeline(this, 'Pipeline', {
+  definition: def,
+  pipeline,
+  pipelineName: 'MyPipeline',
+  vpc,
+  subnetSelection: { subnetType: ec2.SubnetType.PRIVATE },
+  crossAccountKeys: true,
+  buildEnvironment: {
+    image: codebuild.CodeBuildImage.AWS_STANDARD_6,
+    privilegedMode: true,
+  },
+  assetBuildEnvironment: {
+    environmentVariables: {
+      SECRET: { type: codebuild.EnvironmentVariableType.SECRETS_MANAGER, value: 'arn:aws:...:my-secret' },
```

Comment:
Can we use tokens?

Reply:
This would be whatever the CodeBuild L2 API is. If it has a way to do this prettier, then yes, otherwise no.

Reply:
I see - this type is basically coming from the codebuild library. Sounds good.